File: solutions/50/main.cpp
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0) return 1;
        if (x == 0) return 0;  
        double result = 1; 
        // use long long to avoid overflow 
        long long N = n; 
        if (N < 0) {
            x = 1/x;
            N *= -1; 
        } 
        while (N > 0) {
            if (N & 1) {
                N -= 1; 
                result *= x; 
            }
            else {
                N >>= 1; 
                x *= x; 
            }
        } 
        return result; 
    }
};

TEST(Test50, SimpleTest) {
    double x = 2.00000; 
    int n = -2; 
    Solution s; 
    EXPECT_EQ(s.myPow(x, n), 0.25000); 
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS(); 
}

File: solutions/50/main.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0: 
            return 0 
        result = 1 
        if n < 0: 
            x, n = 1/x, -n 
        while n > 0: 
            if n & 1: 
                n -= 1 
                result *= x
            else: 
                n >>= 1 
                x *= x 
        return result 

if __name__ == "__main__": 
    s = Solution()
    x = 2.00000 
    n = 10
    assert s.myPow(x, n) == 1024.00000
    x = 2.00000
    n = -2
    assert s.myPow(x, n) == 0.25000